src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -4040,26 +4259,321 @@ ColocationGroupCreateCommandList(void)
+		appendStringInfo(colocationGroupCreateCommand,
+						 ") SELECT pg_catalog.citus_internal_add_colocation_metadata("
+						 "colocationid, shardcount, replicationfactor, "
+						 "distributioncolumntype, coalesce(c.oid, 0)) "
+						 "FROM colocation_group_data d LEFT JOIN pg_collation c "
+						 "ON (d.distributioncolumncollationname = c.collname "
+						 "AND d.distributioncolumncollationschema::regnamespace"
+						 " = c.collnamespace)");
+
+		SendCommandListToActivatedNodes(context, list_make1(
+											colocationGroupCreateCommand->data));
 	}
+	MemoryContextSwitchTo(oldContext);
 
 	systable_endscan_ordered(scanDescriptor);
 	index_close(colocationIdIndexRel, AccessShareLock);
 	table_close(pgDistColocation, AccessShareLock);
+}
+
 
-	if (!hasColocations)
+/*
+ * SendDependencyCreationCommands sends dependency creation commands to workers
+ * with transactional or nontransactional mode according to transactionMode
+ * inside metadataSyncContext.
+ */
+void
+SendDependencyCreationCommands(MetadataSyncContext *context)
+{
+	MemoryContext oldContext = MemoryContextSwitchTo(context->context);
+
+	/*
+	 * collect all dependencies in creation order and get their ddl commands
+	 */
+	List *dependencies = GetDistributedObjectAddressList();
+
+	/*
+	 * Depending on changes in the environment, such as the enable_metadata_sync guc
+	 * there might be objects in the distributed object address list that should currently
+	 * not be propagated by citus as they are 'not supported'.
+	 */
+	dependencies = FilterObjectAddressListByPredicate(dependencies,
+													  &SupportedDependencyByCitus);
+
+	dependencies = OrderObjectAddressListInDependencyOrder(dependencies);
+
+	MemoryContext commandsContext = AllocSetContextCreate(context->context,
+														  "dependency commands context",
+														  ALLOCSET_DEFAULT_SIZES);
+	MemoryContextSwitchTo(commandsContext);
+	ObjectAddress *dependency = NULL;
+	foreach_ptr(dependency, dependencies)
 	{
-		return NIL;
+		if (IsAnyObjectAddressOwnedByExtension(list_make1(dependency), NULL))
+		{
+			/*
+			 * we expect extension-owned objects to be created as a result
+			 * of the extension being created.
+			 */
+			continue;
+		}
+
+		if (!InMetadataSyncTestMode(context))
+		{
+			MemoryContextReset(commandsContext);
+		}
+
+		/* dependency creation commands */
+		List *ddlCommands = GetAllDependencyCreateDDLCommands(list_make1(dependency));
+
+		if (ddlCommands == NIL)
+		{
+			continue;
+		}
+
+		ddlCommands = lcons(DISABLE_DDL_PROPAGATION,
+							ddlCommands);
+		ddlCommands = lappend(ddlCommands,
+							  ENABLE_DDL_PROPAGATION);
+
+		SendCommandListToActivatedNodes(context, ddlCommands);
+	}
+	MemoryContextSwitchTo(oldContext);
+
+	if (!InMetadataSyncTestMode(context))
+	{
+		MemoryContextDelete(commandsContext);
+		MemoryContextReset(context->context);
+	}
+}
+
+
+/*
+ * SendMetadataCommands sends commands related to pg_dist_shard and,
```

Comment:
Isn't `SendMetadata` too generic? Everything we send here is metadata, this should be `SendMetadataPgDistTableMetadata` or such?